File: Backtracking/SolveSudoku/solve.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool checkInSafeState(vector<vector<char>>& board, int r, int c, char val){
        for(int i=0;i<9;i++){
            if(board[i][c] == val)return false;
        }
        for(int j=0;j<9;j++){
            if(board[r][j] == val)return false;
        }
        int sr = (r/3) * 3;
        int sc = (c/3) * 3;
        for(int i = sr;i < sr+3;i++){
            for(int j = sc;j< sc+3;j++){
                if(board[i][j]==val)return false;
            }
        }
        return true;
    }
    bool solveBack(vector<vector<char>>& board, int r, int c){
        if(r==9)return true;
        else if(c==9)return solveBack(board,r+1,0);
        else if(board[r][c]!='.')return solveBack(board,r,c+1);
        for(int j=0;j<9;j++){
            if(checkInSafeState(board,r,c,char('1' + j))){
                board[r][c] = char('1' + j);
                if(solveBack(board,r,c+1))return true;
                board[r][c] = '.';
            }
        }
        return false;
    }
    void solveSudoku(vector<vector<char>>& board) {
        solveBack(board,0,0);
    }
};

File: Backtracking/SolveSudoku/solve.py
class Solution:
    def checkInSafeState(self, board, r, c, val):
        # Column
        for i in range(9):
            if board[i][c] == val:
                return False

        # Row
        for j in range(9):
            if board[r][j] == val:
                return False

        # 3x3 Box
        sr = (r // 3) * 3
        sc = (c // 3) * 3

        for i in range(sr, sr + 3):
            for j in range(sc, sc + 3):
                if board[i][j] == val:
                    return False

        return True

    def solveBack(self, board, r, c):
        if r == 9:
            return True

        elif c == 9:
            return self.solveBack(board, r + 1, 0)

        elif board[r][c] != '.':
            return self.solveBack(board, r, c + 1)

        for j in range(9):
            val = chr(ord('1') + j)

            if self.checkInSafeState(board, r, c, val):
                board[r][c] = val

                if self.solveBack(board, r, c + 1):
                    return True

                board[r][c] = '.'

        return False

    def solveSudoku(self, board):
        self.solveBack(board, 0, 0)